File: Components/DevExpressVCL/Demos/ExpressEditors Library/Delphi/TokenEditDemo/TokenEditDemoMain.pas
unit TokenEditDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms, Dialogs, BaseForm, Menus, StdCtrls,
  cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters, cxContainer, cxEdit, dxTokenEdit, cxLabel,
  cxGroupBox, cxTextEdit, cxSpinEdit, cxMaskEdit, cxDropDownEdit, cxCheckBox, dxCheckGroupBox, cxCheckComboBox;

type
  TdxTokenEditDemoForm = class(TfmBaseForm)
    cbCloseGlyphPosition: TcxComboBox;
    cbDisplayMask: TcxComboBox;
    cbGlyphPosition: TcxComboBox;
    cbLookupFilterMode: TcxComboBox;
    chbAllowCustomTokens: TcxCheckBox;
    chbConfirmTokenDeletion: TcxCheckBox;
    chbLookupSorted: TcxCheckBox;
    chbPostOnFocusLeave: TcxCheckBox;
    chbReadOnly: TcxCheckBox;
    chcbLookupFilterSources: TcxCheckComboBox;
    chgbLookup: TdxCheckGroupBox;
    dxTokenEdit: TdxTokenEdit;
    gbMain: TcxGroupBox;
    gbOptions: TcxGroupBox;
    gbSample: TcxGroupBox;
    ilSmall: TcxImageList;
    lbCloseGlyphPosition: TcxLabel;
    lbDisplayMask: TcxLabel;
    lbEditValue: TcxLabel;
    lbEditValueDelimiter: TcxLabel;
    lbGlyphPosition: TcxLabel;
    lbInputDelimiters: TcxLabel;
    lbLookupDropDownRows: TcxLabel;
    lbLookupFilterMode: TcxLabel;
    lbLookupFilterSources: TcxLabel;
    lbMaxLineCount: TcxLabel;
    seLookupDropDownRows: TcxSpinEdit;
    seMaxLineCount: TcxSpinEdit;
    teEditValueDelimiter: TcxTextEdit;
    teInputDelimiters: TcxTextEdit;
    procedure cbCloseGlyphPositionPropertiesEditValueChanged(Sender: TObject);
    procedure cbDisplayMaskPropertiesEditValueChanged(Sender: TObject);
    procedure cbGlyphPositionPropertiesEditValueChanged(Sender: TObject);
    procedure cbLookupFilterModePropertiesEditValueChanged(Sender: TObject);
    procedure chbAllowCustomTokensPropertiesEditValueChanged(Sender: TObject);
    procedure chbLookupSortedPropertiesEditValueChanged(Sender: TObject);
    procedure chbPostOnFocusLeavePropertiesEditValueChanged(Sender: TObject);
    procedure chbReadOnlyPropertiesEditValueChanged(Sender: TObject);
    procedure chcbLookupFilterSourcesPropertiesEditValueChanged(Sender: TObject);
    procedure chgbLookupPropertiesEditValueChanged(Sender: TObject);
    procedure dxTokenEditPropertiesEditValueChanged(Sender: TObject);
    procedure dxTokenEditPropertiesTokenClick(Sender: TObject; const ATokenText: string; AToken: TdxTokenEditToken);
    procedure dxTokenEditPropertiesTokenDelete(Sender: TObject; const ATokenText: string; AToken: TdxTokenEditToken;
      var AAllow: Boolean);
    procedure dxTokenEditPropertiesTokenGlyphClick(Sender: TObject; const ATokenText: string;
      AToken: TdxTokenEditToken);
    procedure dxTokenEditPropertiesValidate(Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure seLookupDropDownRowsPropertiesChange(Sender: TObject);
    procedure seMaxLineCountPropertiesChange(Sender: TObject);
    procedure teEditValueDelimiterPropertiesEditValueChanged(Sender: TObject);
    procedure teInputDelimitersPropertiesEditValueChanged(Sender: TObject);
  end;

var
  dxTokenEditDemoForm: TdxTokenEditDemoForm;

implementation

{$R *.dfm}

procedure TdxTokenEditDemoForm.cbCloseGlyphPositionPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.CloseGlyphPosition := TdxTokenEditElementPosition(cbCloseGlyphPosition.ItemIndex);
end;

procedure TdxTokenEditDemoForm.cbDisplayMaskPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.Lookup.DisplayMask := cbDisplayMask.Text;
  if cbDisplayMask.Properties.Items.IndexOf(cbDisplayMask.Text) = -1 then
    cbDisplayMask.Properties.Items.Add(cbDisplayMask.Text);
end;

procedure TdxTokenEditDemoForm.cbGlyphPositionPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.GlyphPosition := TdxTokenEditElementPosition(cbGlyphPosition.ItemIndex);
end;

procedure TdxTokenEditDemoForm.cbLookupFilterModePropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.Lookup.FilterMode := TdxTokenEditLookupFilterMode(cbLookupFilterMode.ItemIndex);
end;

procedure TdxTokenEditDemoForm.chbAllowCustomTokensPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.AllowAddCustomTokens := chbAllowCustomTokens.Checked;
  dxTokenEdit.ValidateEdit;
end;

procedure TdxTokenEditDemoForm.chbLookupSortedPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.Lookup.Sorted := chbLookupSorted.Checked;
end;

procedure TdxTokenEditDemoForm.chbPostOnFocusLeavePropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.PostEditValueOnFocusLeave := chbPostOnFocusLeave.Checked;
end;

procedure TdxTokenEditDemoForm.chbReadOnlyPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.ReadOnly := chbReadOnly.Checked;
end;

procedure TdxTokenEditDemoForm.chcbLookupFilterSourcesPropertiesEditValueChanged(Sender: TObject);
var
  AFilterSources: TdxTokenEditLookupFilterSources;
begin
  AFilterSources := [];
  if chcbLookupFilterSources.States[0] = cbsChecked then
    Include(AFilterSources, tefsText);
  if chcbLookupFilterSources.States[1] = cbsChecked then
    Include(AFilterSources, tefsDisplayText);
  dxTokenEdit.Properties.Lookup.FilterSources := AFilterSources;
end;

procedure TdxTokenEditDemoForm.chgbLookupPropertiesEditValueChanged(Sender: TObject);
begin
  dxTokenEdit.Properties.Lookup.Active := chgbLookup.CheckBox.Checked;
end;

procedure TdxTokenEditDemoForm.dxTokenEditPropertiesEditValueChanged(Sender: TObject);
var
  ACaption: string;
begin
  ACaption := 'Edit Value: ';
  if dxTokenEdit.Text <> '' then
    ACaption := ACaption + dxTokenEdit.Text
  else
    ACaption := ACaption + '(empty)';
  lbEditValue.Caption := ACaption;
end;

procedure TdxTokenEditDemoForm.dxTokenEditPropertiesTokenClick(Sender: TObject; const ATokenText: string;
  AToken: TdxTokenEditToken);
begin
  MessageDlg('Clicked token: ' + ATokenText, mtInformation, [mbOK], 0);
end;

procedure TdxTokenEditDemoForm.dxTokenEditPropertiesTokenDelete(Sender: TObject; const ATokenText: string;
  AToken: TdxTokenEditToken; var AAllow: Boolean);
begin
  AAllow := not chbConfirmTokenDeletion.Checked or
    (MessageDlg('Do you want to delete token: ' + ATokenText + '?', mtConfirmation, [mbYes, mbNo], 0) = mrYes);
end;

procedure TdxTokenEditDemoForm.dxTokenEditPropertiesTokenGlyphClick(Sender: TObject; const ATokenText: string;
  AToken: TdxTokenEditToken);
begin
  MessageDlg('Clicked glyph of token: ' + ATokenText, mtInformation, [mbOK], 0);
end;

procedure TdxTokenEditDemoForm.dxTokenEditPropertiesValidate(Sender: TObject; var DisplayValue: Variant;
  var ErrorText: TCaption; var Error: Boolean);
var
  AError: Boolean;
begin
  AError := (VarToStr(DisplayValue) <> '') and not chbAllowCustomTokens.Checked and
    (dxTokenEdit.Properties.Tokens.FindByText(DisplayValue) = nil);
  Error := Error or AError;
  if AError then
  begin
    if ErrorText <> '' then
      ErrorText := ErrorText + #13#10;
    ErrorText := ErrorText + 'Custom tokens are not allowed: ' + DisplayValue;
  end;
end;

procedure TdxTokenEditDemoForm.seLookupDropDownRowsPropertiesChange(Sender: TObject);
begin
  if (seLookupDropDownRows.Value <> Null) and (seLookupDropDownRows.Value >= 0) then
    dxTokenEdit.Properties.Lookup.DropDownRows := seLookupDropDownRows.Value
  else
    seLookupDropDownRows.Value := 0;
end;

procedure TdxTokenEditDemoForm.seMaxLineCountPropertiesChange(Sender: TObject);
begin
  if (seMaxLineCount.Value <> Null) and (seMaxLineCount.Value >= 0) then
    dxTokenEdit.Properties.MaxLineCount := seMaxLineCount.Value
  else
    seMaxLineCount.Value := 0;
end;

procedure TdxTokenEditDemoForm.teEditValueDelimiterPropertiesEditValueChanged(Sender: TObject);
begin
  if teEditValueDelimiter.Text <> '' then
    dxTokenEdit.Properties.EditValueDelimiter := teEditValueDelimiter.Text[1]
  else
    teEditValueDelimiter.Text := dxTokenEditDefaultEditValueDelimiter;
  dxTokenEdit.ValidateEdit;
end;

procedure TdxTokenEditDemoForm.teInputDelimitersPropertiesEditValueChanged(Sender: TObject);
begin
  if teInputDelimiters.Text <> '' then
    dxTokenEdit.Properties.InputDelimiters := teInputDelimiters.Text
  else
    teInputDelimiters.Text := dxTokenEditDefaultInputDelimiters;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressEditors Library/CBuilder/TokenEditDemo/TokenEditDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "TokenEditDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BaseForm"
#pragma link "cxCheckBox"
#pragma link "cxCheckComboBox"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxDropDownEdit"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxMaskEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTextEdit"
#pragma link "dxCheckGroupBox"
#pragma link "dxTokenEdit"
#pragma resource "*.dfm"
TdxTokenEditDemoForm *dxTokenEditDemoForm;
//---------------------------------------------------------------------------
__fastcall TdxTokenEditDemoForm::TdxTokenEditDemoForm(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TdxTokenEditDemoForm::cbCloseGlyphPositionPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->CloseGlyphPosition = TdxTokenEditElementPosition(cbCloseGlyphPosition->ItemIndex);
}
//---------------------------------------------------------------------------
void __fastcall TdxTokenEditDemoForm::cbDisplayMaskPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->Lookup->DisplayMask = cbDisplayMask->Text;
  if(cbDisplayMask->Properties->Items->IndexOf(cbDisplayMask->Text) == -1)
    cbDisplayMask->Properties->Items->Add(cbDisplayMask->Text);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::cbGlyphPositionPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->GlyphPosition = TdxTokenEditElementPosition(cbGlyphPosition->ItemIndex);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::cbLookupFilterModePropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->Lookup->FilterMode = TdxTokenEditLookupFilterMode(cbLookupFilterMode->ItemIndex);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chbAllowCustomTokensPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->AllowAddCustomTokens = chbAllowCustomTokens->Checked;
  dxTokenEdit->ValidateEdit();
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chbLookupSortedPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->Lookup->Sorted = chbLookupSorted->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chbPostOnFocusLeavePropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->PostEditValueOnFocusLeave  = chbPostOnFocusLeave->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chbReadOnlyPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->ReadOnly = chbReadOnly->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chcbLookupFilterSourcesPropertiesEditValueChanged(TObject *Sender)
{
  TdxTokenEditLookupFilterSources AFilterSources;

  AFilterSources.Clear();
  if(chcbLookupFilterSources->States[0] == cbsChecked)
	AFilterSources << tefsText;
  if(chcbLookupFilterSources->States[1] == cbsChecked)
	AFilterSources << tefsDisplayText;
  dxTokenEdit->Properties->Lookup->FilterSources = AFilterSources;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::chgbLookupPropertiesEditValueChanged(TObject *Sender)
{
  dxTokenEdit->Properties->Lookup->Active = chgbLookup->CheckBox->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::dxTokenEditPropertiesEditValueChanged(TObject *Sender)
{
  UnicodeString ACaption;

  ACaption = "Edit Value: ";
  if(dxTokenEdit->Text != "")
	ACaption += dxTokenEdit->Text;
  else
	ACaption += "(empty)";
  lbEditValue->Caption = ACaption;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::dxTokenEditPropertiesTokenClick(TObject *Sender,
          const UnicodeString ATokenText, TdxTokenEditToken *AToken)
{
  MessageDlg("Clicked token: " + ATokenText, mtInformation, TMsgDlgButtons() << mbOK, 0);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::dxTokenEditPropertiesTokenDelete(TObject *Sender,
          const UnicodeString ATokenText, TdxTokenEditToken *AToken, bool &AAllow)
{
  AAllow = !chbConfirmTokenDeletion->Checked ||
	(MessageDlg("Do you want to delete token: " + ATokenText + "?", mtConfirmation, TMsgDlgButtons() << mbYes << mbNo, 0) == mrYes);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::dxTokenEditPropertiesTokenGlyphClick(TObject *Sender,
          const UnicodeString ATokenText, TdxTokenEditToken *AToken)
{
  MessageDlg("Clicked glyph of token: " + ATokenText, mtInformation, TMsgDlgButtons() << mbOK, 0);
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::dxTokenEditPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  bool AError;

  AError = (VarToStr(DisplayValue) != "") && !chbAllowCustomTokens->Checked &&
  	(dxTokenEdit->Properties->Tokens->FindByText(DisplayValue) == (0x0));
  Error = Error || AError;
  if(AError)
  {
	if(ErrorText != "")
	  ErrorText = ErrorText + '\n';
	ErrorText = ErrorText + "Custom tokens are not allowed: " + DisplayValue;
  }
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::seLookupDropDownRowsPropertiesChange(TObject *Sender)
{
  if((seLookupDropDownRows->Value != Null()) && (seLookupDropDownRows->Value >= 0))
	dxTokenEdit->Properties->Lookup->DropDownRows = seLookupDropDownRows->Value;
  else
	seLookupDropDownRows->Value = 0;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::seMaxLineCountPropertiesChange(TObject *Sender)
{
  if((seMaxLineCount->Value != Null()) && (seMaxLineCount->Value >= 0))
	dxTokenEdit->Properties->MaxLineCount = seMaxLineCount->Value;
  else
	seMaxLineCount->Value = 0;
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::teEditValueDelimiterPropertiesEditValueChanged(TObject *Sender)
{
  if(teEditValueDelimiter->Text != "")
	dxTokenEdit->Properties->EditValueDelimiter = teEditValueDelimiter->Text[1];
  else
	teEditValueDelimiter->Text = dxTokenEditDefaultEditValueDelimiter;
  dxTokenEdit->ValidateEdit();
}
//---------------------------------------------------------------------------

void __fastcall TdxTokenEditDemoForm::teInputDelimitersPropertiesEditValueChanged(TObject *Sender)
{
  if(teInputDelimiters->Text != "")
	dxTokenEdit->Properties->InputDelimiters = teInputDelimiters->Text;
  else
	teInputDelimiters->Text = dxTokenEditDefaultInputDelimiters;
}
//---------------------------------------------------------------------------

